### PR TITLE
add early stop

### PR DESCRIPTION
**Pull request by ceci3**
1. add median stop

example:
```python
from paddleslim.nas import SANAS
from paddleslim.nas.early_stop import MedianStop


steps = 10
epochs = 7

config = [('MobileNetV2Space')]
sanas = SANAS(config, server_addr=("", 8732), save_checkpoint=None)

### add for early stop
earlystop = MedianStop(sanas, 2)

for step in range(steps):
    archs = sanas.next_archs()[0]
    for epoch in range(epochs):
        for data in train_reader():
            loss = archs(data)

        for data in test_reader():
            loss = archs(data)
        avg_cost = np.mean(loss)

        ### add for early stop
        status = earlystop.get_status(step, avg_cost, epochs)
        if status == 'BAD':
            break;

    sanas.reward(avg_cost)
```

### CODE REVIEW

**Comment by wanghaoshuang** [comment]
```
for step in range(steps):
    archs = sanas.next_archs()[0]
    for epoch in range(epochs):
        for data in train_reader():
            loss = archs(data)
```
你这个示例里的`loss = archs(data)`是什么意思？是动态图？还是只是一种简写用以表示训练过程？

File: docs/zh_cn/api_cn/early_stop.rst [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+ 获取当前实验当前result的状态。
+
+ **参数：**
+ - **step<int>** - 当前实验是当前client中的第几个实验。
+ - **result<float>** - 当前实验的中间步骤的result，可以为损失值，也可以为准确率等指标，只要和`mode`对应即可。
+ - **epochs<int>** - 在搜索过程中每个实验需要运行的总得epoch数量。
+
+ **返回：**
+ 返回当前实验在当前epoch的状态，为`GOOD`或者`BAD`，如果为`BAD`，则代表当前实验可以早停。
```
Comment (wanghaoshuang) [inline docs/zh_cn/api_cn/early_stop.rst:43]:
```
if early_stop.stop():
    break
```
用户接口这么设计会更好么？当前用`get_status`是参考了其它框架是实现么？还是有什么通用的准则？

File: docs/zh_cn/api_cn/early_stop.rst [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ .. code-block:: python
+   from paddleslim.nas import SANAS
+   from paddleslim.nas.early_stop import MedianStop
+   
+   steps = 10
+   epochs = 7
+   
+   config = [('MobileNetV2Space')]
+   sanas = SANAS(config, server_addr=("", 8732), save_checkpoint=None)
+   earlystop = MedianStop(sanas, 2)
```
Comment (wanghaoshuang) [inline docs/zh_cn/api_cn/early_stop.rst:56]: 有没有可能把earlystop作为SANAS的一个参数？然后在`sanas.next_archs()`里调用`earlystop.get_status`?

File: paddleslim/nas/early_stop/median_stop/median_stop.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+_logger = get_logger(__name__, level=logging.INFO)
+
+completed_history = dict()
+
+
+def return_completed_history():
+    return completed_history
+
+
+class MedianStop(EarlyStopBase):
```
Comment (wanghaoshuang) [inline paddleslim/nas/early_stop/median_stop/median_stop.py:33]: 我们有必要支持用户自定义early stop策略么？

**Comment by ceci3** [comment]
> ```
> for step in range(steps):
>     archs = sanas.next_archs()[0]
>     for epoch in range(epochs):
>         for data in train_reader():
>             loss = archs(data)
> ```
> 
> 你这个示例里的`loss = archs(data)`是什么意思？是动态图？还是只是一种简写用以表示训练过程？

这儿只是一个简写表示训练过程

File: docs/zh_cn/api_cn/early_stop.rst (hunk 1 above)
Comment (ceci3) [reply docs/zh_cn/api_cn/early_stop.rst:43]: get_status是参考了其他的框架的实现，表示当前的一个状态，万一用户只是想知道当前的实验的状态不想停掉，用stop感觉容易有歧义

File: docs/zh_cn/api_cn/early_stop.rst (hunk 2 above)
Comment (ceci3) [reply docs/zh_cn/api_cn/early_stop.rst:56]: 按照现在SANAS的实现和用法不太可能把early stop集成进去。next_archs()表示每个token的整体情况，这是要在每个训练epoch之后都要拿到一个可以对比的值，感觉不太可能这样调用。我之前想过要集成进去，但是这样就需要server端了解每一轮实验细节，而且也需要模型代码的配合，目前我们的实现server端没有办法知道实验细节。

File: paddleslim/nas/early_stop/median_stop/median_stop.py (hunk 3 above)
Comment (ceci3) [reply paddleslim/nas/early_stop/median_stop/median_stop.py:33]: 按照现在的实现其实没有必要支持...

File: docs/zh_cn/api_cn/early_stop.rst (hunk 2 above)
Comment (wanghaoshuang) [reply docs/zh_cn/api_cn/early_stop.rst:56]: 哦，我理解错了。

**Review by wanghaoshuang — APPROVED** [review]
LGTM
